### ISSUE TITLE
fping - strange dips in ping every 2m30s (4 vs 5)

### ISSUE THREAD

**Issue by luckman212** [netdata/netdata#8148]
I just installed Netdata v1.19.0-539-nightly + fping plugin, pretty standard install on Ubuntu 19.10 running in VMware. Everything is working great 👍

I noticed these strange "dips" roughly every 2.5min on the graphs, where the # of pings sent / received changes from 5 down to 4.  "returned" pct% remains at 100, but the graphs make it appear as if something is wrong. Not sure what the reason for this could be? Some timing issue?

Here's my fping config, which is essentially unchanged from defaults: (5 pings every 1s)
```
fping="/usr/local/bin/fping"
hosts="host1 host2 host3..."
update_every=1
ping_every="200"
fping_opts="-R -b 56 -i 1 -r 0 -t 5000"
```
screenshot:
![image](https://user-images.githubusercontent.com/1992842/74987684-c41e9980-5409-11ea-82f4-fa129eeee163.png)

**Comment by luckman212** [comment]
**Update:** to try to rule out other external factors, I migrated the VM to a more powerful host, increased the allocated RAM to 8GB, CPU to 4x, switched the vNIC to vmxnet type, updated the base OS, and connected the host to a different physical switch.

The dips/blips have unfortunately persisted. Not sure where to go to troubleshoot this.

**Comment by luckman212** [comment]
**Update 2:** tried a full (clean) reinstall using Debian 10 instead, to see if it was possibly an underlying OS issue. Problem persists there as well. Tomorrow maybe I'll try Arch or CentOS, but I don't think it's the OS.

**Comment by vlvkobal** [comment]
It is not a Netdata issue. I get similar dips, but they are produced by `fping` itself. The strange thing is that when `fping` uses the netdata protocol, `xmt = 5`, `rcv = 4`, but in the standard format `xmt/rcv/%loss = 4/4/0%`.

![image](https://user-images.githubusercontent.com/22008543/75437570-13f7e600-595f-11ea-8c11-14f695b8738c.png)

`/usr/local/bin/fping -N -l -Q 1 -p 200 -R -b 56 -i 1 -r 0 -t 5000 google.com`
```
BEGIN fping.google_com_packets
SET xmt = 5
SET rcv = 5
END
BEGIN fping.google_com_quality
SET returned = 100
END
BEGIN fping.google_com_latency
SET min = 2793
SET avg = 2798
SET max = 2801
END
BEGIN fping.google_com_packets
SET xmt = 5
SET rcv = 4
END
BEGIN fping.google_com_quality
SET returned = 100
END
BEGIN fping.google_com_latency
SET min = 2781
SET avg = 2787
SET max = 2794
END
...
BEGIN fping.google_com_packets
SET xmt = 5
SET rcv = 4
END
BEGIN fping.google_com_quality
SET returned = 100
END
BEGIN fping.google_com_latency
SET min = 2789
SET avg = 2794
SET max = 2800
END
BEGIN fping.google_com_packets
SET xmt = 4
SET rcv = 4
END
BEGIN fping.google_com_quality
SET returned = 100
END
BEGIN fping.google_com_latency
SET min = 2790
SET avg = 2794
SET max = 2804
END
BEGIN fping.google_com_packets
SET xmt = 5
SET rcv = 5
END
BEGIN fping.google_com_quality
SET returned = 100
END
BEGIN fping.google_com_latency
SET min = 2786
SET avg = 2795
SET max = 2803
END
```
`/usr/local/bin/fping -l -Q 1 -p 200 -R -b 56 -i 1 -r 0 -t 5000 google.com`
```
[12:50:35]
google.com : xmt/rcv/%loss = 5/5/0%, min/avg/max = 27.7/27.8/27.9
[12:50:36]
google.com : xmt/rcv/%loss = 4/4/0%, min/avg/max = 27.7/27.8/27.9
...
google.com : xmt/rcv/%loss = 4/4/0%, min/avg/max = 27.8/27.8/27.9
[12:51:01]
google.com : xmt/rcv/%loss = 5/5/0%, min/avg/max = 27.7/27.8/27.9

```

**Comment by vlvkobal** [comment]
You can slightly decrease the value of the `-p` option to get rid of the dips.

**Comment by luckman212** [comment]
@vlvkobal Thanks for checking. Should I open an issue at [schweikert/fping](https://github.com/schweikert/fping) ? I had a quick peek and it seems like maybe this is related to fping's internal timer resolution of 10us vs 1us - there's [issue#101](https://github.com/schweikert/fping/issues/101) and a workaround in [PR#136](https://github.com/schweikert/fping/pull/136) (hasn't been merged yet though)

**Comment by vlvkobal** [comment]
> Should I open an issue at schweikert/fping ?

Yes, please. From my side, I've already opened a PR (https://github.com/schweikert/fping/pull/172) to fix `xmt/rcv` difference in the `fping` output for Netdata.

**Comment by luckman212** [comment]
I compiled the [high-resolution version](https://github.com/tycho/fping/tree/high-resolution-clock-sources) and replaced it on my test system. Sadly it not fix this issue.

For now, is there a way to hide the `fping.{hostname}_packets)` portion of the graphs just to avoid distraction?

**Comment by tycho** [comment]
I fixed the netdata-related code in the high resolution fping branch. The units had changed so the results netdata saw were not making proper sense. Can you give that branch another try, @luckman212?

**Comment by luckman212** [comment]
@tycho I've compiled the updated branch, it's working fine for me 👍 
Do I need to adjust the `fping_opts=` line in `/etc/netdata/fping.conf` to enable the 1us resolution or is that default?

**Comment by luckman212** [comment]
> You can check your existing config [..] Add to the `netdata.conf`
> ```
> [fping.google_com_packets]
> 	enabled = no
> ```

Thank you @vlvkobal ! But, I have about 30 hosts, it's too much work!

I decided to edit `fping.c` and comment out the part that generates the packet count instead...
![image](https://user-images.githubusercontent.com/1992842/75511876-94e5ca80-59bd-11ea-946b-7ceff45607b8.png)

Working nicely!
![image](https://user-images.githubusercontent.com/1992842/75512100-57ce0800-59be-11ea-9a89-d04cc59e74e5.png)

**Comment by tycho** [comment]
> @tycho I've compiled the updated branch, it's working fine for me 👍
> Do I need to adjust the `fping_opts=` line in `/etc/netdata/fping.conf` to enable the 1us resolution or is that default?

No tweaking needed, it's the default (and only) way to measure latency in that branch. It's possible to tweak how it does so at compile-time (e.g. which clock source it uses) but the defaults are best.
